tools/slo-generator/slo_generator/backends/elasticsearch.py

```diff
@@ -17,74 +17,124 @@
 """
 
 import logging
-import pprint
-from slo_generator.backends.base import MetricBackend
+
 from elasticsearch import Elasticsearch
 
+from slo_generator.backends.base import MetricBackend
+
 LOGGER = logging.getLogger(__name__)
 
 
 class ElasticsearchBackend(MetricBackend):
-    """Backend for querying metrics from Elasticsearch."""
+    """Backend for querying metrics from ElasticSearch."""
 
     def __init__(self, **kwargs):
         self.client = kwargs.get('client')
         if self.client is None:
             self.client = Elasticsearch(**kwargs)
 
-    def query(self, window, filter):
-        raise NotImplementedError()
+    def good_bad_ratio(self, **kwargs):
+        """Query two timeseries, one containing 'good' events, one containing
+        'bad' events.
+
+        Args:
+            timestamp (int): UNIX timestamp.
+            window (int): Window size (in seconds).
+            kwargs (dict): Extra arguments needed by this computation method.
+                project_id (str): GCP project id to fetch metrics from.
+                measurement (dict): Measurement config.
+                    filter_good (str): Query filter for 'good' events.
+                    filter_bad (str): Query filter for 'bad' events.
+
+        Returns:
+            tuple: A tuple (good_event_count, bad_event_count)
+        """
+        window = kwargs['window']
+        index = kwargs['measurement']['index']
```

Comment:
When we use `get`, the argument will default to `None` when not specified, which effectively makes it an optional argument. Keying is used when the argument is required, so it would stop execution if a required field is needed.

More generally, we should do config parsing and check arguments for a config type in the `__init__` method. Also keeping this for the next major version.

Reply:
See https://github.com/ocervell/professional-services/tree/slo-generator-v1.0.0

Reply:
I'd encourage you to *not* split work into "major" versions. Reviewing a big PR like this is much more work than reviewing several small PRs.

Reply:
I agree ! Being the only developer and getting so many feature requests for this it's not that easy. Will try to break features into PRs for the future